Fix waze_travel_time component ERROR on startup

VirtualL opened the pull request:
## Description:
Fix the unhandled exception with Waze Travel Time sensor upon startup,
by adding Throttle before update_interval are starting.

**Related issue (if applicable):**
fixes #16676

Error treated:
```
Error doing job: Future exception was never retrieved
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/src/app/homeassistant/components/sensor/waze_travel_time.py", line 74, in <lambda>
    EVENT_HOMEASSISTANT_START, lambda _: sensor.update())
  File "/usr/src/app/homeassistant/components/sensor/waze_travel_time.py", line 204, in update
    routes = params.calc_all_routes_info(real_time=self._realtime)
  File "/usr/local/lib/python3.6/site-packages/WazeRouteCalculator/WazeRouteCalculator.py", line 145, in calc_all_routes_info
    routes = self.get_route(npaths, time_delta)
  File "/usr/local/lib/python3.6/site-packages/WazeRouteCalculator/WazeRouteCalculator.py", line 95, in get_route
    response_json = response.json()
  File "/usr/local/lib/python3.6/site-packages/requests/models.py", line 897, in json
    return complexjson.loads(self.text, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/simplejson/__init__.py", line 518, in loads
    return _default_decoder.decode(s)
  File "/usr/local/lib/python3.6/site-packages/simplejson/decoder.py", line 370, in decode
    obj, end = self.raw_decode(s)
  File "/usr/local/lib/python3.6/site-packages/simplejson/decoder.py", line 400, in raw_decode
    return self.scan_once(s, idx=_w(s, idx).end())
simplejson.errors.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```
## Example entry for `configuration.yaml` (if applicable):
```yaml
# Example entry for configuration.yaml
sensor:
  - platform: waze_travel_time
    origin: Montréal, QC
    destination: Québec, QC
    region: 'US'
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

~~If user exposed functionality or configuration variables are added/changed:~~
 ~~- [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)~~

If the code communicates with devices, web services, or third-party tools:
~~- [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).~~
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
~~- [ ] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.~~
~~- [ ] New files were added to `.coveragerc`.~~

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works- Tested on Docker Container 0.85.0 , 0.85.1 , 0.86.0.dev0

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

VirtualL committed:
Fix waze_travel_time component ERROR on startup

Fix the unhandled exception with Waze Travel Time sensor upon startup,
by adding Throttle before update_interval are starting.

homeassistant commented:
Hi @VirtualL, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#20316).

Once you do that we will be able to review and accept this pull request.

Thanks!

VirtualL committed:
add missing whitespace after ','

VirtualL committed:
fix line too long (80 > 79 characters)

VirtualL committed:
lint

MartinHjelmare commented on homeassistant/components/sensor/waze_travel_time.py:
```diff
@@ -51,6 +52,8 @@
     vol.Optional(CONF_INCL_FILTER): cv.string,
     vol.Optional(CONF_EXCL_FILTER): cv.string,
     vol.Optional(CONF_REALTIME, default=DEFAULT_REALTIME): cv.boolean,
+    vol.Optional(CONF_UPDATE_INTERVAL, default=timedelta(minutes=5)):
```
This is already present in the base platform config schema.

MartinHjelmare commented on homeassistant/components/sensor/waze_travel_time.py:
```diff
@@ -94,6 +99,7 @@ def __init__(self, name, origin, destination, region,
         self._state = None
         self._origin_entity_id = None
         self._destination_entity_id = None
+        self.update = Throttle(interval)(self.update)
```
Just add the throttle as a decorator to `update` with fixed throttling time.

VirtualL replied:
Thank you for the feedback, I was Fix the interval 
Can you clarify a little more about this one?

MartinHjelmare replied:
We should set the minimum interval between updates to a safe fixed time interval. That doesn't need to be configurable.
```py
@Throttle(MIN_TIME_BETWEEN_UPDATES)
def update(self):
   ...
```

Note that a throttle on the entity update won't hinder other entities from making requests to the api within this minimum time interval. We're not throttling the whole platform just per entity. Is this ok?

VirtualL replied:
Excellent explanation, updated the code, thank you

MartinHjelmare reviewed (changes requested):
Revert all the changes. Then add the throttle as decorator with fixed interval.

MartinHjelmare commented:
https://github.com/home-assistant/home-assistant/issues/14527 is unrelated to this fix and has already been fixed. I've updated the description.

VirtualL committed:
fix interval to use const

MartinHjelmare commented on homeassistant/components/sensor/waze_travel_time.py:
```diff
@@ -51,6 +51,8 @@
     vol.Optional(CONF_INCL_FILTER): cv.string,
     vol.Optional(CONF_EXCL_FILTER): cv.string,
     vol.Optional(CONF_REALTIME, default=DEFAULT_REALTIME): cv.boolean,
+    vol.Optional(CONF_SCAN_INTERVAL, default=timedelta(seconds=300)):
```
Remove this option. It's already present in the base platform config schema.

MartinHjelmare commented on homeassistant/components/sensor/waze_travel_time.py:
```diff
@@ -39,7 +40,6 @@
 
 REGIONS = ['US', 'NA', 'EU', 'IL', 'AU']
 
-SCAN_INTERVAL = timedelta(minutes=5)
```
Add this back. This sets the default update interval for the platform.

VirtualL committed:
Change to Throttle as a decorator to update

Change to Throttle as a decorator to update instead of self.update = Throttle(interval)(self.update)
remove unnecessary code.

VirtualL commented:
Is there anyone who can help check why not passes the tests?

MartinHjelmare commented:
```
./homeassistant/components/sensor/waze_travel_time.py:89:17: E128 continuation line under-indented for visual indent
homeassistant.components.sensor.waze_travel_time
homeassistant/components/sensor/waze_travel_time.py:89:0: C0330: Wrong continued indentation (add 1 space).
                incl_filter, excl_filter, realtime):
                ^| (bad-continuation)
homeassistant/components/sensor/waze_travel_time.py:9:0: C0411: standard import "import logging" should be placed before "from homeassistant.util import Throttle" (wrong-import-order)
homeassistant/components/sensor/waze_travel_time.py:11:0: C0411: third party import "import voluptuous as vol" should be placed before "from homeassistant.util import Throttle" (wrong-import-order)
homeassistant/components/sensor/waze_travel_time.py:13:0: C0412: Imports from package homeassistant are not grouped (ungrouped-imports)
```

VirtualL committed:
fix  indentations

VirtualL committed:
Update waze_travel_time.py

VirtualL committed:
Update waze_travel_time.py

MartinHjelmare commented on homeassistant/components/sensor/waze_travel_time.py:
```diff
@@ -10,13 +10,15 @@
-from homeassistant.const import (
-    ATTR_ATTRIBUTION, CONF_NAME, CONF_REGION, EVENT_HOMEASSISTANT_START,
-    ATTR_LATITUDE, ATTR_LONGITUDE)
+from homeassistant.const import (ATTR_ATTRIBUTION, CONF_NAME,
```
Please revert this formatting change.

MartinHjelmare commented on homeassistant/components/sensor/waze_travel_time.py:
```diff
@@ -10,13 +10,15 @@
 import homeassistant.helpers.config_validation as cv
 from homeassistant.helpers import location
 from homeassistant.helpers.entity import Entity
 
```
Remove the blank line here.

VirtualL committed:
Update waze_travel_time.py

MartinHjelmare reviewed (approved):
Thanks!

VirtualL commented:
thank you for the guidance! next time will be much easier 🥇 .

IIIdefconIII commented:
which release contains this fix im on 0.87 now

MartinHjelmare commented:
None yet. I'll tag it with 0.87.1.

MartinHjelmare commented on homeassistant/components/sensor/waze_travel_time.py:
```diff
@@ -67,7 +69,7 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
                             incl_filter, excl_filter, realtime)
 
-    add_entities([sensor])
+    add_entities([sensor], True)
```
We shouldn't have included this change. We want to wait with updating until home assistant start event. We already have a listener for that, which will call update.

IIIdefconIII replied:
so fix will be in 0.87.2?

MartinHjelmare replied:
Please don't comment here. Let's keep the conversation in the issue.
